aips/pepper-service-for-keyless-accounts.md

```diff
@@ -0,0 +1,236 @@
+
+## Alternative solutions
+
+### Let dApp end users to manage their own pepper?
+The whole point of using keyless accounts is to save users from memorizing extra secrets,
+and this breaks the point.
+
+### Let dApp service owners to manage the pepper?
+It would make keyless account dApp-dependent:
+if the dependency dApp dies, all the dependent accounts are not accessible.
+
+## Specification
+
+### Preliminaries
+
+A few data items involved in the pepper service interaction is briefly explained below.
+For the full glossaries, see [keyless account specification](https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-61.md#specification).
+
+A pair of *ephemeral private key (ESK) and ephemeral public key (EPK)*
+is generated each time the user logs in to the OIDC provider in order to use a keyless flow-enabled app.
+A keyless transaction requires an ephemeral signature generated using the ESK.
+
+An *EPK expiry date* is a timestamp specified for every EPK. If an EPK's expiry date is past, the EPK is considered expired and cannot be used to sign transactions.
+
+*EPK blinder* are some random bytes used to blind EPK and its expiry date
+when committing them into the `nonce` field of the sign request sent to the OIDC provider.
+(The issued `JWT` will contained the same `nonce`.)
+
+### Public parameters
+
+#### Nonce derivation
+A function `derive_nonce()` is required to commit user's EPK, EPK blinder and the expiry timestamp into a nonce.
+The nonce is then sent to the OIDC provider and signed as part of JWT payload.
+
+The same function needs to be implemented as [part of the ZK relation](https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-61.md#snark-friendly-hash-functions),
+so SNARK-friendly hash function should be preferred.
+The reference implementation uses Poseidon hash function over BN254.
+
+#### Verifiable unpredictable function (VUF)
+Roughly speaking, a VUF is a mapping `f(x)` implicitly determined by a key pair `sk, pk` such that:,
+- for a random input `x`, it is hard to predict `f(x)` without private key `sk`;
+- for an output `y` claimed to be `f(x)`, anyone can verify if the claim is true using the public key `pk`.
+
+These properties make the VUF output a good candidate for the pepper used in the keyless account flow.
+
+Some concrete VUF schemes applicable here:
+- BLS VUF (over BLS12-381 G1 group, refered to as `Bls12381G1Bls` below).
+- [Pinkas VUF](https://eprint.iacr.org/2024/198.pdf).
+
+Below we refer to the VUF scheme being used in pepper calculation as `vuf`,
+and the evaluation function as `vuf.eval`.
+
+#### Asymmetric encryption
+
+In addition, an asymmetric encryption with variable-length input is required to encrypt the pepper with end user's ESK,
+to ensure only the end user can see the pepper.
+Without encryption, it is still possible for the OIDC provider to learn the account address by interacting with the pepper service on behalf of the end user.
```

Comment:
hmm, encryption actually doesn't prevent the provider from getting the pepper, since the provider can generate the ephemeral key pairs by themselves.

It's really just for the leaked JWT cases...

correct? @alinush

Reply:
Correct!